### PR TITLE
lambdify: Included scipy backend automatically and added relevant functions

### LINKED ISSUE #7229: Add scipy to lambdify
```
SciPy has versions of several functions that are in SymPy that aren't in NumPy, such as special functions ( http://docs.scipy.org/doc/scipy/reference/special.html ).  We need to add a scipy translation dictionary to lambdify.py, and include "scipy" automatically just as we do numpy.
```

Original issue for #7229: http://code.google.com/p/sympy/issues/detail?id=4130
Original author: https://code.google.com/u/asmeurer@gmail.com/

### LINKED ISSUE #14973: "Not supported by Python" from lambdify
Reported at https://github.com/sympy/sympy/pull/14971#issuecomment-407830406

For example

```py
>>> f = lambdify(x, rf(18,x))
>>> print(inspect.getsource(f))
def _lambdifygenerated(x):
    return (  # Not supported in Python:
  # RisingFactorial
RisingFactorial(18, x))
```

The problem is that `rf` is defined in the `MPMATH_TRANSLATIONS`, which the printer doesn't know about. 

This doesn't affect behavior at all, it's just confusing when you look at the source code. 

CC @bjodah @cbm755

### LINKED ISSUE #14975: SciPy lambdify printer doesn't know about rf
The scipy rf is `poch`. The lambdify printer doesn't know about it. 

This probably requires fixing https://github.com/sympy/sympy/issues/7229 first. 

See also https://github.com/sympy/sympy/pull/14971

### PR DESCRIPTION
**Pull request by ramvenkat98**
#### References to other Issues or PRs

Fixes #7229 
Fixes #14973 
Fixes #14975 
Referred to PR 10319

#### Brief description of what is fixed or changed

Added relevant scipy.special functions to sympy/sympy/printing/pycode.py
Added use of SciPy printer to lambdify
Added fix so that lambdify automatically uses scipy if available
Added tests to sympy/utilities/tests/test_lambdify.py
Fixed error in sympy/sympy/printing/pycode.py, replacing lgamma with
loggamma in the _not_in_numpy list

#### Release Notes

* utilities
  * lambdify uses scipy automatically if available

### CODE REVIEW

**Review by bjodah — COMMENTED** [review]
Looks good to me. Perhaps we should set numpy.random.seed

File: sympy/utilities/lambdify.py [hunk 1]
```diff
@@ -360,14 +364,19 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
...
+            except ImportError:
+                # Use either numpy (if available) or python.math where possible.
+                # XXX: This leads to different behaviour on different systems and
+                #      might be the reason for irreproducible errors.
+                modules = ["math", "mpmath", "sympy"]
+            else:
+                modules = ["numpy"]
         else:
-            modules = ["numpy"]
+            modules = ["scipy"]
```
Comment (asmeurer) [inline sympy/utilities/lambdify.py]: Shouldn't it be `["scipy", "numpy"]`?

**Comment by asmeurer** [comment]
You need to write out "fixes" separately for each issue in the PR description so they will all be automatically closed.

**Comment by ramvenkat98** [comment]
@asmeurer @bjodah 

I've updated the pull request according to the review:
1) Set numpy.random.seed(0) before the random tests
2) Also adding "numpy" to the modules list in lambdify when scipy is present (and changing the order of Printer selection accordingly to look for SciPy printer before NumPy printer)
3) Wrote out fixes separately for each issue in the PR

However, exactly one Travis test is failing right now and it seems to be a random failure unrelated to the fix- could this test be re-run?

Thanks

**Comment by bjodah** [comment]
Yes, I restarted the test, thank you for telling.

**Comment by ramvenkat98** [comment]
@asmeurer @bjodah 

Hi,

are these fixes ok and can the pull request be merged?

Thanks again

**Comment by ramvenkat98** [comment]
@asmeurer @bjodah 

Hi,

may I have an update on the status of this pull request? Should I fix anything more or is it ok to be merged? (Sorry for commenting multiple times in a row, I figured it's been a few days since the last activity so I brought it up again).

Thanks

**Comment by bjodah** [comment]
@ramvenkat98 thank you very much for your contribution (and sorry for the wait)

**Comment by GProtoZeroW** [comment]
@ramvenkat98 
I just tried a simple call on of the Bessel function of the first kind in sympy and tried to lambdify it to scipy's special.jv and it did not work. Is there something that I am missing to invoke the translation from sympy to scipy in the improved lambdfy in sympy1.3

```
from sympy import *
import scipy.special as SciSpec

#scipy ref 
#https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.jv.html#scipy.special.jv
Order=1; Arg=1.2
SciSpec.jv(Order, Arg)
>>>0.49828905756721553

#sympy call
nu, z=symbols('nu, z')
#https://docs.sympy.org/latest/modules/functions/special.html?highlight=bessel#module-
sympy.functions.special.bessel
b=besselj(nu, z); b
N(b.subs({nu:Order, z:Arg}))
>>>0.498289057567215

bN=lambdify((nu, z), b, dummify=False)
bN(Order, Arg)
>>>NameError                                 Traceback (most recent call last)
<ipython-input-10-3f7a3bb1d5f0> in <module>()
      1 bN=lambdify((nu, z), b, dummify=False)
----> 2 bN(Order, Arg)

~/anaconda3/lib/python3.6/site-packages/numpy/__init__.py in <lambda>(nu, z)

NameError: name 'besselj' is not defined
```

**Comment by ramvenkat98** [comment]
@GProtoZeroW @asmeurer @bjodah 

When using the version of sympy pulled from Github, I don't get this error; however, when using sympy installed as the Anaconda package, I do. Perhaps, this is because the issue has already been fixed on Github but did not make it to the currently released version?

**Comment by moorepants** [comment]
SymPy 1.1.1 is available from the default channel for Anaconda. If you want SymPy >=1.2 you need to install from the conda-forge channel (at least until the default get updated).

**Comment by GProtoZeroW** [comment]
@ramvenkat98 I did the Anaconda install I will have to do the github one to test out. And thanks for adding scipy to lambdfiy it's been needed. Just two questions: one does `Heaviside` from sympy now translate over to numpy's  `heaviside` (https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.heaviside.html). And two do you have a Ex of using the conversion from a sympy integral to a scipy integral; cause that would be something quite useful

**Comment by ramvenkat98** [comment]
@GProtoZeroW I'm not sure of the answers to those questions - you could ask them on the sympy google group, or make feature requests for them - to my knowledge, Heaviside hasn't been translated from sympy to numpy but it certainly seems useful.

**Pull request merged**